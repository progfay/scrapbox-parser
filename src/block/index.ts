import { convertToCodeBlock } from "./CodeBlock";
import { convertToLine } from "./Line";
import { convertToTable } from "./Table";
import { convertToTitle } from "./Title";

import type { CodeBlock } from "./CodeBlock";
import type { Line } from "./Line";
import type { Pack } from "./Pack";
import type { Table } from "./Table";
import type { Title } from "./Title";

/**
 * Scrapbox block type
 */
export type Block = Title | CodeBlock | Table | Line;

export const convertToBlock = (pack: Pack): Block => {
	switch (pack.type) {
		case "title":
			return convertToTitle(pack);

		case "codeBlock":
			return convertToCodeBlock(pack);

		case "table":
			return convertToTable(pack);

		case "line":
			return convertToLine(pack);
	}
};
