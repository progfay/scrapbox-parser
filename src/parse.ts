import { BlockType, convertToBlock } from './block'
import { BlockComponentType, convertToBlockComponents } from './block/BlockComponent'
import { PackedBlockComponentType, packBlockComponents } from './block/PackedBlockComponent'

type PageType = {
  title: string
  blocks: BlockType[]
}

export const convertToBlocks = (blockComponents: BlockComponentType[]): BlockType[] => {
  const packedBlockComponents: PackedBlockComponentType[] = packBlockComponents(blockComponents)
  return packedBlockComponents.map(convertToBlock)
}

export const parse = (input: string): PageType => {
  const blockComponents: BlockComponentType[] = convertToBlockComponents(input.trim())

  const [firstBlock, ...body] = blockComponents
  const title = firstBlock && firstBlock.text ? firstBlock.text : 'Untitled'
  const blocks = convertToBlocks(body)

  return { title, blocks }
}
