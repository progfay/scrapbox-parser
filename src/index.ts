import { parse } from './parse'

export default parse
export { BlockType } from './block'
export { CodeBlockType } from './block/CodeBlock'
export { TableType } from './block/Table'
export { LineType } from './block/Line'
export { LineNodeType } from './block/node'
export { QuoteNodeType } from './block/node/QuoteNode'
export { HelpfeelNodeType } from './block/node/HelpfeelNode'
export { StrongImageNodeType } from './block/node/StrongImageNode'
export { StrongIconNodeType } from './block/node/StrongIconNode'
export { StrongNodeType } from './block/node/StrongNode'
export { FormulaNodeType } from './block/node/FormulaNode'
export { DecorationNodeType, DecorationType } from './block/node/DecorationNode'
export { CodeNodeType } from './block/node/CodeNode'
export { BlankNodeType } from './block/node/BlankNode'
export { UrlNodeType } from './block/node/UrlNode'
export { GoogleMapNodeType } from './block/node/GoogleMapNode'
export { InternalLinkNodeType } from './block/node/InternalLinkNode'
export { IconNodeType } from './block/node/IconNode'
export { HashTagNodeType } from './block/node/HashTagNode'
export { PlainNodeType } from './block/node/PlainNode'
