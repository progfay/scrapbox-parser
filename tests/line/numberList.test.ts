import { describe, expect, it } from "vitest";
import { parse } from "../../src";

describe("numberList", () => {
	it("Minimum numberList", () => {
		expect(
			parse("1. ", {
				hasTitle: false,
			}),
		).toMatchSnapshot();
	});

	it("Simple numberList", () => {
		expect(
			parse("1. Simple numberList", {
				hasTitle: false,
			}),
		).toMatchSnapshot();
	});

	it("1. with no space is not numberList", () => {
		expect(
			parse("1.not numberList", {
				hasTitle: false,
			}),
		).toMatchSnapshot();
	});

	it("No head 1. is not numberList", () => {
		expect(
			parse("a 1. not numberList", {
				hasTitle: false,
			}),
		).toMatchSnapshot();
	});

	it("Quoted 1. is not numberList", () => {
		expect(parse("> 1. Quoted", { hasTitle: false })).toMatchSnapshot();
	});
});
