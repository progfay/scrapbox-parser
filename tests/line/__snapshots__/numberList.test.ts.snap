// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`numberList > 1. with code 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [],
        "number": 1,
        "raw": "1. ",
        "rawNumber": "1",
        "type": "numberList",
      },
      {
        "raw": "\`code\`",
        "text": "code",
        "type": "code",
      },
    ],
    "type": "line",
  },
]
`;

exports[`numberList > 1. with decoration 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [],
        "number": 1,
        "raw": "1. ",
        "rawNumber": "1",
        "type": "numberList",
      },
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "raw": "deco",
            "text": "deco",
            "type": "plain",
          },
        ],
        "raw": "[* deco]",
        "rawDecos": "*",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
]
`;

exports[`numberList > 1. with no space is not numberList 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "1.not numberList",
        "text": "1.not numberList",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`numberList > Minimum numberList 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [],
        "number": 1,
        "raw": "1. ",
        "rawNumber": "1",
        "type": "numberList",
      },
    ],
    "type": "line",
  },
]
`;

exports[`numberList > No head 1. is not numberList 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "a 1. not numberList",
        "text": "a 1. not numberList",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`numberList > Quoted 1. is not numberList 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": " 1. Quoted",
            "text": " 1. Quoted",
            "type": "plain",
          },
        ],
        "raw": "> 1. Quoted",
        "type": "quote",
      },
    ],
    "type": "line",
  },
]
`;

exports[`numberList > Simple numberList 1`] = `
[
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Simple numberList",
            "text": "Simple numberList",
            "type": "plain",
          },
        ],
        "number": 1,
        "raw": "1. Simple numberList",
        "rawNumber": "1",
        "type": "numberList",
      },
    ],
    "type": "line",
  },
]
`;
